compiler/verif/z3refinement.ml

```diff
@@ -1079,47 +1029,48 @@ and prove_function ctx n local_env arg_list typenv =
     Use function space to determine if argument list has expected type
     from function space
 *)
-    (* if (Hashtbl.mem !stream_space n) then ()
-    (* if it is a stream *)
-
-    else ( *)(
       if (Hashtbl.mem !function_space n) 
         (* refinement function, make sure input list obeys constraints *)
         then (
           let ref_fun = Hashtbl.find !function_space n in
-          debug (Printf.sprintf "TODO -- check if arguments obey constraints\n");
-          print_function_temp n ref_fun;
+          debug (Printf.sprintf "checking if arguments obey constraints\n");
+          print_function n ref_fun;
           print_env local_env;
           (* let expr_test = vc_gen_expression ctx local_env (List.hd arg_list) None in
           Printf.printf "Arg_list[0]: %s\n" (Expr.to_string expr_test); *)
           let constraint_env = { exp_env = ref ref_fun.argument_constraints ; var_env = ref_fun.creation_env.var_env } in 
           let arguments = List.map (fun elem -> create_z3_var ctx constraint_env elem) ref_fun.argument_list in
-          Printf.printf "map2 error is here!!\n";
           Printf.printf "name of function:%s\n" n;
           Printf.printf "length of ref_fun.argument_list:%d\n" (List.length ref_fun.argument_list);
-          Printf.printf "# of Arguments in arg_list:%d arguments:%d\n" (List.length arg_list) (List.length arguments);
-          let checks = List.map2 (fun elem1 elem2 -> create_validation_check ctx constraint_env elem1 elem2) arg_list arguments in
+          let user_args = List.map (fun elem -> ( 
+            if (is_ETuple elem) then (
+              vc_gen_tuple_list ctx constraint_env elem None
+            ) else (
+              [vc_gen_expression ctx constraint_env elem None]
+            )
+            )) arg_list in
+          let user_args_flat = List.flatten user_args in
+          check_function_input user_args_flat arguments;
+          (* let checks = list_map3 (fun elem1 elem2 elem3 -> create_validation_check ctx constraint_env elem1 elem2 elem3) user_args_flat arguments ref_fun.argument_constraints in *)
+          let checks = List.map (fun elem -> Expr.substitute elem arguments user_args_flat) ref_fun.argument_constraints in
+          ignore(List.map (fun elem -> debug(Expr.to_string elem)) checks);
           (* let environment_constraints = List.map (get_environment_constraints ctx local_env typenv) arg_list in *)
           (* print_env ({ exp_env = ref( checks @ environment_constraints); var_env = Hashtbl.create 0}); *)
-          let check_env = { exp_env = ref (checks @ !(local_env.exp_env)); var_env = Hashtbl.create 0} in
-          check_validity ctx constraint_env check_env;
+          let check_env = checks @ !(local_env.exp_env) @ !(ref_fun.creation_env.exp_env) in
+          check_validity ctx checks;
```

Comment:
We noticed a bug here during our code review. `check_validity` returns true even if constraints are not satisfied due to the use of an implication statement. Quick fix is to replace the original implication `x -> x` with `T -> x` (just true in the antecedent) to keep with the original function interface with minimal changes.

Reply:
Fixed with the latest commit